### PR TITLE
Forward `**kwargs` in `__init_subclass__` to super

### LINKED ISSUE #10732: `DeclarativeBase` doesn't forward keyword arguments in `__init_subclass__`
### Describe the bug

`DeclarativeBase.__init_subclass__` doesn't forward unknown keyword arguments meaning `Base` **MUST** come last when using a Mixin that uses `__init_subclass__` with keyword arguments.

The docs don't reflect this:

> There’s no fixed convention for the order of mixins and base classes. Normal Python method resolution rules apply, and the above example would work just as well with:
> 
> ```python
> class MyModel(Base, HasLogRecord, CommonMixin):
>     name: Mapped[str] = mapped_column()
> ```

This should be a simple fix for `DeclarativeBase` but opens a wider discussion to do this for other or all classes that use `__init_subclass__` e.g. `MappedAsDataclass`?

Happy to open a PR for this, draft without tests here https://github.com/sqlalchemy/sqlalchemy/pull/10733.
If accepted where should a test for this live?

### Optional link from https://docs.sqlalchemy.org which documents the behavior that is expected

https://docs.sqlalchemy.org/en/20/orm/declarative_mixins.html#composing-mapped-hierarchies-with-mixins

### SQLAlchemy Version in Use

2.0.23

### DBAPI (i.e. the database driver)

psycopg2

### Database Vendor and Major Version

PostgreSQL 15

### Python Version

3.11

### Operating system

OSX

### To Reproduce

```python
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column


class Base(DeclarativeBase):
    pass


class Mixin:
    def __init_subclass__(cls, arg1: str, **kwargs: object) -> None:
        super().__init_subclass__(**kwargs)


# Works because `Base` is last
class TableOne(Mixin, Base, arg1=""):
    __tablename__ = "table_one"
    pk: Mapped[int] = mapped_column(primary_key=True)


# Doesn't work because `Base` is first and doesn't forward **kwargs
class TableTwo(Base, Mixin, arg1=""):
    __tablename__ = "table_two"
    pk: Mapped[int] = mapped_column(primary_key=True)

```

### Error

```
Traceback (most recent call last):
  File "example.py", line 18, in <module>
    class TableTwo(Base, Mixin, arg1=""):
TypeError: __init_subclass__() got an unexpected keyword argument 'arg1'
``` 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by michaeloliverx**
Fixes: #10732

### Checklist

This pull request is:

- [ ] A documentation / typographical / small typing error fix
	- Good to go, no issue or tests are needed
- [x] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

File: lib/sqlalchemy/orm/decl_api.py [hunk 1]
```diff
@@ -839,13 +839,13 @@ def _sa_inspect_instance(self) -> InstanceState[Self]:
         def __init__(self, **kw: Any):
             ...
 
-    def __init_subclass__(cls) -> None:
+    def __init_subclass__(cls, **kwargs: Any) -> None:
```
Comment (zzzeek) [inline lib/sqlalchemy/orm/decl_api.py:842]: prefer **kw

**Review by zzzeek — CHANGES_REQUESTED** [review]
please add tests at https://github.com/sqlalchemy/sqlalchemy/blob/d8cab0694e6dc9f7b7fd02a4b5a8ae6bbb2e896e/test/orm/declarative/test_basic.py#L900 thanks!

**Comment by CaselIT** [comment]
thanks for the update. The test could use combinations, but let's bring it in gerrit

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 0a76ab159e941543ca587936928ad25beb2e3e86 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change 0a76ab159e941543ca587936928ad25beb2e3e86: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4998

**Comment by sqla-tester** [comment]
**Michael Bayer** (zzzeek) wrote:

pretty good.

to move this along, run pre-commit and also add a changelog file.  otherwise we will try to get to it...

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4998

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 7fdeec1f3224f48213c9c9af5f3e7e5d0904dafa of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset 7fdeec1f3224f48213c9c9af5f3e7e5d0904dafa added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4998

**Comment by CaselIT** [comment]
> to move this along, run pre-commit and also add a changelog file. otherwise we will try to get to it...

added to gerrit

**Comment by sqla-tester** [comment]
**Federico Caselli** (CaselIT) wrote:

added changelog

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4998

**Review by sqla-tester — COMMENTED** [review]
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4998

* lib/sqlalchemy/dialects/mysql/base.py (line 1574): not totally sure how this got here , but if we dont fix the typo here I'll just forget about it

**Review by sqla-tester — COMMENTED** [review]
**Federico Caselli** (CaselIT) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4998

* lib/sqlalchemy/dialects/mysql/base.py (line 1574): Oh I think I had it locally since I fixed on 2.0 here https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5025 and I guess it got committed while pushing the changelog

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4998 has been **merged**. Congratulations! :)

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5009 has been **merged**. Congratulations! :)

**Pull request closed without merging**